### PR TITLE
Fix exploded edge filter on persistent graph

### PR DESCRIPTION
**Pull request by ljeub-pometry**
### What changes were proposed in this pull request?

- Filtering an edge update on a persistent graph turns it into a deletion to keep the semantics sensible
- Refactor the edge filtering implementation to make this possible

### Why are the changes needed?

- Exploded edge filtering on a persistent graph had weird effects on the exploded edges (the time-range where remaining edges are valid got extended by the removals)

### How was this patch tested?

more tests

### CODE REVIEW

File: raphtory/src/db/api/view/internal/filter_ops.rs [hunk 1]
```diff
@@ -71,13 +148,97 @@ impl<G: BoxableGraphView + Clone> FilterOps for G {
...
+        self.internal_node_list_trusted()
+            && self.node_filter_includes_edge_filter()
+            && self.node_filter_includes_edge_layer_filter()
+            && self.node_filter_includes_exploded_edge_filter()
+    }
+
+    fn filter_edge(&self, edge: EdgeStorageRef) -> bool {
+        self.internal_filter_edge(edge, self.layer_ids()) && self.filter_edge_from_nodes(edge) && {
+            let time_semantics = self.edge_time_semantics();
+            edge.layer_ids_iter(self.layer_ids()).any(|layer_id| {
```
Comment (miratepuffin) [inline raphtory/src/db/api/view/internal/filter_ops.rs:179]: Possibly need to add some tests where we have some node filters, some edge filters and some exploded edge filters

File: raphtory/src/db/api/view/internal/time_semantics/event_semantics.rs [hunk 2]
```diff
@@ -190,18 +190,53 @@ impl NodeTimeSemanticsOps for EventSemantics {
...
-        layer_ids: &LayerIds,
+        layer_id: usize,
+        w: Range<i64>,
+    ) -> bool {
+        edge.filtered_additions(layer_id, &view).active_t(w.clone())
+            || edge.filtered_deletions(layer_id, &view).active_t(w)
+    }
+
+    fn include_exploded_edge<G: GraphView>(&self, eid: ELID, t: TimeIndexEntry, view: G) -> bool {
+        view.layer_ids().contains(&eid.layer())
```
Comment (miratepuffin) [inline raphtory/src/db/api/view/internal/time_semantics/event_semantics.rs:219]: Do we need this?

File: raphtory/src/db/api/view/internal/time_semantics/filtered_edge.rs [hunk 3]
```diff
@@ -33,13 +47,13 @@ impl<'a, 'graph: 'a, G: GraphViewOps<'graph>> TimeIndexOps<'a>
 
     #[inline]
     fn active(&self, w: Range<Self::IndexType>) -> bool {
-        if self.view.edge_history_filtered() {
+        if self.view.internal_exploded_edge_filtered() {
```
Comment (miratepuffin) [inline raphtory/src/db/api/view/internal/time_semantics/filtered_edge.rs:50]: Can we think up some tests for making sure this is always providing the correct answer

File: raphtory/src/db/api/view/internal/time_semantics/filtered_edge.rs [hunk 4]
```diff
@@ -56,42 +70,117 @@ impl<'a, 'graph: 'a, G: GraphViewOps<'graph>> TimeIndexOps<'a>
...
 }
 
+#[derive(Clone)]
+pub struct InvertedFilteredEdgeTimeIndex<'graph, G> {
+    eid: ELID,
+    time_index: TimeIndexRef<'graph>,
+    view: G,
+}
+
+impl<'a, 'graph: 'a, G: GraphViewOps<'graph>> TimeIndexOps<'a>
```
Comment (miratepuffin) [inline raphtory/src/db/api/view/internal/time_semantics/filtered_edge.rs:116]: This doesn't seem to be doing the right thing anymore

File: raphtory/src/db/api/view/internal/time_semantics/persistent_semantics.rs [hunk 5]
```diff
@@ -468,10 +561,10 @@ impl EdgeTimeSemanticsOps for PersistentSemantics {
     fn edge_layers<'graph, G: GraphViewOps<'graph>>(
         self,
         e: EdgeStorageRef<'graph>,
-        view: G,
+        _view: G,
```
Comment (miratepuffin) [inline raphtory/src/db/api/view/internal/time_semantics/persistent_semantics.rs:564]: Whats this about

File: raphtory/src/db/api/view/internal/time_semantics/persistent_semantics.rs [hunk 6]
```diff
@@ -617,14 +712,25 @@ impl EdgeTimeSemanticsOps for PersistentSemantics {
         view: G,
         w: Range<i64>,
     ) -> Option<i64> {
-        if edge_alive_at_end(e, w.end, &view) {
-            Some(w.end - 1)
+        let interior_window = w.start.saturating_add(1)..w.end;
```
Comment (miratepuffin) [inline raphtory/src/db/api/view/internal/time_semantics/persistent_semantics.rs:715]: Can we add a test where we add an edge at time 5 then look at the graph at time 10, what is the earliest and latest time of the graph - does this make sense

File: raphtory/src/db/api/view/internal/time_semantics/persistent_semantics.rs [hunk 7]
```diff
@@ -932,7 +1051,7 @@ impl EdgeTimeSemanticsOps for PersistentSemantics {
         prop_id: usize,
         t: TimeIndexEntry,
     ) -> Option<Prop> {
-        EventSemantics.temporal_edge_prop_last_at(e, view, prop_id, t)
+        EventSemantics.temporal_edge_prop_last_at(e, view, prop_id, t) // TODO: double check this
```
Comment (miratepuffin) [inline raphtory/src/db/api/view/internal/time_semantics/persistent_semantics.rs:1054]: Do the TODO

File: raphtory/src/db/api/view/internal/time_semantics/persistent_semantics.rs [hunk 8]
```diff
@@ -993,7 +1113,9 @@ impl EdgeTimeSemanticsOps for PersistentSemantics {
     ) -> impl Iterator<Item = (TimeIndexEntry, usize, Prop)> + Send + Sync + 'graph {
         e.filtered_temporal_prop_iter(prop_id, view.clone(), layer_ids)
             .map(|(layer, props)| {
-                let deletions = e.filtered_deletions(layer, &view);
+                let deletions = e
```
Comment (miratepuffin) [inline raphtory/src/db/api/view/internal/time_semantics/persistent_semantics.rs:1116]: Can we add some more tests on persistent graphs for properties after deletions - do we like the result

File: raphtory/src/db/graph/views/valid_graph.rs [hunk 9]
```diff
@@ -153,6 +136,38 @@ mod tests {
...
+        expected.resolve_layer(Some("a")).unwrap();
+        assert_graph_equal(&gv, &expected);
+        let gvm = gv.materialize().unwrap();
+        assert_graph_equal(&gv, &gvm);
+    }
+
+    #[test]
+    fn test_single_deleted_edge_persistent() {
+        let g = PersistentGraph::new();
+        g.delete_edge(0, 0, 0, Some("a")).unwrap();
```
Comment (miratepuffin) [inline raphtory/src/db/graph/views/valid_graph.rs:158]: Can we add a test for combining exploded_edge_filter with a valid check

**Review by miratepuffin — APPROVED** [review]
Couple of comments to turn into issues, but otherwise LGTM

* We need to work out if we want to seperate out edges where all updates have been converted into deletions
